### DISCUSSION TITLE
Next Position Cursor Prediction (Inspired from Cursor Obvioslly...)

### DISCUSSION THREAD

**Discussion by PLAZMAMA** [olimorris/codecompanion.nvim#323] [Ideas]
I would like to take a crack at creating a cursor position prediction feature kind off like what the cursor team is working on.
Would this fit the projects goals/direction @olimorris?

**Comment by olimorris** [comment]
It's a very interesting feature.

I've shied away from implementing a suggestions strategy for a couple of reasons. Firstly, I don't think I could improve upon `copilot.vim`. Secondly, the amount of context that's needed to do suggestions properly, means that it's really only Ollama models that are viable for most users from a cost perspective. I expect that the next cursor position would need to know the history of where the user's cursor has been, the buffer's code and probably the overall structure of the cwd.

If you think this could be accomplished differently, let me know because I do think this is hugely compelling as a feature.

**Comment by PLAZMAMA** [comment]
## Specific Responses to Your Comment _(Thanks for responding btw)_

1. #### Firstly, I don't think I could improve upon copilot.vim
2. 
Good point, `copilot`, `supermaven` and `codeium` do it so good already so I found myself just using them. But I think we should state this clearly in the readme to avoid confusion (which I had while trying to figure out where the inline suggestion feature was and how to activate it)

3. #### Secondly, the amount of context that's needed to do suggestions properly, means that it's really only Ollama models that are viable for most users from a cost perspective.
4. 
I think this can be addressed in the two following ways:
-  Make context window size adjustable to the users need via a `context_window_size` option
- Limit context only be file specific.

## My Current Implementation Strategy

Taking into account everything you laid out and my vision for this feature, I will formulate an implementation strategy below. I would like more feedback on this feature, especially stuff like the last comment.

### Implementation overview (will be written in a bit as I have to switch to my laptop to commute)

I think the implementation is composed of four parts
- Context
- Prompt
- Response parsing and prediction display
- Event/prediction trigger

#### Context
I think the best way to get the context efficiently will be the following:
- Keep a list of the history of edited line numbers for each file. This list/table will look like the following:
``` lua
local edits_history = {
    {
        ["name"] = "file_name_1.lua"
        ["edits_line_num_history"] = {
            33,
            22,
            33,
            9,
            102,
            ...
        }
    },
     {
        ["name"] = "file_name_2.lua"
        ["edits_line_num_history"] = {
            43,
            24,
            43,
            111,
            12,
            ...
        }
    },
    ...
}
```
- **I'm going to flush everything below next week as i have school deadlines and a busy weekend**
- **I will tag you (Ollimorris) once I'm finished**

#### Prompt

#### Response parsing and prediction display

#### Event/prediction trigger

**Reply by olimorris** [reply]
I think we'll also need to capture the contents on the line number as well. As people edit the buffer, line 12, 5 cursor moves ago may no longer be the same. This gives the context to the LLM of what the user is doing and how they're modifying the code. I also think it's important to think about the sequencing of the moves as well. I may be moving between `file_name_2` and line 12 and `file_name_1` and line 102 frequently, how will the LLM know about that? I think a table like:

```lua
{
 { name = "lua/myapp/file_name_2.lua", line = 12, ft = "lua", content = [[print("hello world")]]},
 { name = "lua/myapp/file_name_1.lua", line = 102, ft = "lua", content = [[vim.iter(tbl)]]},
}
```

possibly makes more sense.

**Reply by olimorris** [reply]
One thing you could do is turn this into a tool initially, and see how the LLM performs. You won't even need to fork the repository to do this as you can load local files in the config. Also, the `codecompanion.utils.buffers` file may have some useful methods to leverage.

**Reply by PLAZMAMA** [reply]
First of all, I want to say thank you for responding so quickly and with such good insights.

> I think we'll also need to capture the contents on the line number as well. As people edit the buffer, line 12, 5 cursor moves ago may no longer be the same. This gives the context to the LLM of what the user is doing and how they're modifying the code. I also think it's important to think about the sequencing of the moves as well. I may be moving between `file_name_2` and line 12 and `file_name_1` and line 102 frequently, how will the LLM know about that? I think a table like:
> 
> ```lua
> {
>  { name = "lua/myapp/file_name_2.lua", line = 12, ft = "lua", content = [[print("hello world")]]},
>  { name = "lua/myapp/file_name_1.lua", line = 102, ft = "lua", content = [[vim.iter(tbl)]]},
> }
> ```
> 
> possibly makes more sense.

I agree, but I want to expand this concept event further.

> One thing you could do is turn this into a tool initially, and see how the LLM performs. You won't even need to fork the repository to do this as you can load local files in the config. Also, the `codecompanion.utils.buffers` file may have some useful methods to leverage.

Yes, I've been thinking of doing it like that just as a "prototype" to see how it performs. Ive been also thinking of maybe turning this into its own plugin that runs on a local network that you fine tune as you gather more data locally. This is because I think this extension focuses on for of the chatting side of AI (aka explaining, git commit message generation, boilerplate generation etc...)

**Reply by PLAZMAMA** [reply]
Would you like me to close the discussion or keep it open?

**Reply by olimorris** [reply]
> Would you like me to close the discussion or keep it open?

Completely up to you. Here to bounce ideas off if helpful. Very excited by this!

**Reply by PLAZMAMA** [reply]
Thanks, I will keep this discussion up to update you on my progress and maybe ask for advice, I know Ill need some.

So let me start with the first update I guess...
I plan to make this idea into its own extension with a fine-tuned llm specifically for this that the user can fine-tune further locally on their pc. This LLM will be able to run on most CPUs fast thanks to Microsoft's [BitNet](https://github.com/microsoft/BitNet) project. However, initially I will test/experiment with pre-built llms like lama3 and so on...

Right now I'm super busy trying to graduate school ASAP (3-ish classes left), after I graduate I plan on spending 6 weeks straight making this project. So stay tuned :)

**Reply by olimorris** [reply]
I'm not sure if by extension you mean plugin but I would strongly urge you to consider turning this into its own plugin at some point. I think it's potentially too good to merge into mine as so many other users could benefit from it. It really could be [Harpoon](https://github.com/ThePrimeagen/harpoon) but on steroids. From a CodeCompanion.nvim perspective, if there's any integration that we'd need to do, I'd make it a top priority.

BitNet looks like a great suggestion too.

Good luck with grad school and hope you nail it!

**Reply by PLAZMAMA** [reply]
> I'm not sure if by extension you mean plugin.

Yes, I meant plugin, my bad. VScode and Chrome are too ingrained in my head, unfortunately. 

> I would strongly urge you to consider turning this into its own plugin at some point. I think it's potentially too good to merge into mine as so many other users could benefit from it.

Exactly!!!, I already created the empty repo for it https://github.com/PLAZMAMA/bunnyhop.nvim. 

> It really could be [Harpoon](https://github.com/ThePrimeagen/harpoon) but on steroids.

Yes, that is one of my favorite Nvim extensions and what initially made me switch from VSCode to Neovim, it also inspired this approach to the idea of cursor prediction.

> Good luck with grad school and hope you nail it!

Thanks! But I think there was a little misunderstanding I'm finishing my Bachelor's in CS rn, I wish I was already doing my Master's lol.
I went with a less conventional route of learning code in High School -> 2yrs coding work for $ -> use $ to do a Bachelor's.

> From a CodeCompanion.nvim perspective, if there's any integration that we'd need to do, I'd make it a top priority.

I have some ideas for that as well... but I'm trying not to jump too far ahead and focus on getting the base functionality in a good state first before I bring these ideas up and get too excited.

**Reply by olimorris** [reply]
Hey @PLAZMAMA, hope you're good. Did you make any progress on this btw?

**Reply by PLAZMAMA** [reply]
Hi @olimorris, since we last spoke I've completed my last 2 classes and am now finishing my capstone project. I hope to start working on this by the end of the month.

**Reply by olimorris** [reply]
> Hi @olimorris, since we last spoke I've completed my last 2 classes and am now finishing my capstone project. I hope to start working on this by the end of the month.

Good luck on the project!

**Comment by PLAZMAMA** [comment]
Hi @olimorris, I've been working on the extension for a few days now, and so far, it's going alright. I've been able to create something that when a keybinding is pressed **TRIES** to predict your next cursor location and jump to it. I've still got a long way before this is usable though and there are a lot of improvements to be made.

I've run into some problems that I haven't been able to figure out yet so if you have any tips it would be appreciated.
- I've been using `vim.fn.system` for running curl which works but isn't async. When I try `vim.system` for curl commands I get a nested braces in the URL error. I've tried to debug it by stepping through Neovim's API code and wasn't able to figure out what is the raw command it creates. Is there a way to "dry run" this function or be able to see what is the final command it formulates?
- I've tried figuring out how to send requests to copilot chat, I haven't seen any documentation on the internet and wasn't able to figure it out from reading your and CopilotChat.nvim. Do you know of any examples or contained code I can use to accomplish the auth, token handling, and requests?

Additionally, after trying to debug all of these things for the last couple of days, I think it would be really helpful for people to have a library they can use to query all of the different kinds of AI API's. This will prevent the need to rewrite the same adapter logic over and over again for every plugin the wants to utilize these API's. Since Code Companion has the biggest adapter library, I think it might be worth for someone to try to take a look at the adapters implementation you got here and try to convert it to a single independent library for people to use.

Lastly, I wanted to say thank you again for nudging me to make this plugin happen and I hope soon I'll be able to say that its at a usable state.

**Reply by olimorris** [reply]
Hey @PLAZMAMA great to hear from you and this sounds awesome!

> Is there a way to "dry run" this function or be able to see what is the final command it formulates?

I can't say I use `vim.fn.system` that much so unsure of its nuances. I'm using Plenary.Curl for all of CodeCompanion because of its asynchronous properties and because it's so well battle tested at this point. It can handle streaming too. Also, a great tip, if you use CodeCompanion's `/fetch` slash command and send the raw file to an LLM, you can ask it in detail how it works. That's helped me out loads when building the plugin.

> Do you know of any examples or contained code I can use to accomplish the auth, token handling, and requests?

Check out my implementation, [here](https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/adapters/copilot.lua). We've solved all of that. To see the full payload that's sent to the LLM, you may need to set the debug level to "TRACE" in your CodeCompanion settings and examine the JSON file that the plugin creates

Really excited to try this out.

**Comment by PLAZMAMA** [comment]
@olimorris I've been working on the plugin([bunnyhop.nvim](https://github.com/PLAZMAMA/bunnyhop.nvim)) since we last talked and it is now at a point where I feel it can be of some use. Its still in a very alpha stage but you can start using it if you want. I have a lot of improvements (such as the context creation, tests, auto release, etc..) which I will continue making of course and if you have any other suggestions it would be appreciated.

**Reply by olimorris** [reply]
Amazing work!!! I can't wait to try it out. Reading the code base, I need to supply a hugging face API key in my config?

**Reply by PLAZMAMA** [reply]
Good question, you need to create a hugging face environment variable with your API key and supply the name of that environment variable in the config. I'll clarify it in the readme.

**Reply by olimorris** [reply]
I've now got the plugin setup so will use it over the next 24 hours and report back 👍🏼

**Reply by PLAZMAMA** [reply]
I've added the ability to try different hugging face models since hugging face removes and adds models from the API so try playing around with the models if the default doesn't work.

**Reply by olimorris** [reply]
So far when I try to trigger the plugin, it just sends my cursor to the top of the buffer. Is there a way to debug my settings?

My config:

```lua
  {
    "PLAZMAMA/bunnyhop.nvim",
    lazy = false,
    keys = {
      {
        "<M-h>",
        function()
          require("bunnyhop").hop()
        end,
        desc = "[H]op to predicted location.",
      },
    },
    opts = { api_key = "HF_API_KEY" },
  },
```

**Reply by PLAZMAMA** [reply]
You're config is correct, the issue was with Hugging Face. They removed the model I was using so I switched it to a different model which didn't output in the correct format. They brought the original model back up now, so if you update you plugin now, it should work now because I reverted the model back to its original. Also, I added the ability to choose which model you want to use. You can look at hugging face's [models page](https://huggingface.co/models?inference=warm&sort=trending) and choose a different model if the reverted one doesn't work. An example config would look like so:

```lua
 {
    "PLAZMAMA/bunnyhop.nvim",
    lazy = false,
    keys = {
      {
        "<M-h>",
        function()
          require("bunnyhop").hop()
        end,
        desc = "[H]op to predicted location.",
      },
    },
    opts = { api_key = "HF_API_KEY", model="Qwen/Qwen2.5-Coder-32B-Instruct" }, -- This is the default model (which to me works the best)
    -- you can put any hugging face model by their name which
    -- is following the format {company name}/{model name}, 
    -- ex. meta-llama/Llama-3.3-70B-Instruct, mistralai/Mistral-Nemo-Instruct-2407, etc...
  },
```

Also, I'm now working towards making the Github copilot adapter because hugging face has some annoying reliability issues as described above.

**Comment by Million-mo** [comment]
Is there an update on cursor prediction?

**Reply by olimorris** [reply]
Yep - https://github.com/PLAZMAMA/bunnyhop.nvim

**Reply by PLAZMAMA** [reply]
FYI, I'm currently working on tracking and improving the accuracy of the predictions. 
There is a lot to improve but I'm working on it.

**Reply by Million-mo** [reply]
Based on cursor's work, I'm thinking about the need for cursor prediction, as if cursor wants to work more on next action prediction.

**Reply by PLAZMAMA** [reply]
I don't know what they want to work on more as I'm not on their team. I just used their ideas as inspiration for my own workflow improvements. I do think next action prediction is a cool idea though, Its definitely worth exploring! I can see how something like cursor prediction can be a part of that as well.
